File: apps/desktop_flutter/lib/app/core/errors/app_error.dart
class AppError implements Exception {
  AppError(this.message, {this.code, this.statusCode});

  final String message;
  final String? code;
  final int? statusCode;

  factory AppError.notFound(String resource) => AppError(
        '$resource not found',
        code: 'NOT_FOUND',
        statusCode: 404,
      );

  factory AppError.badRequest(String message) => AppError(
        message,
        code: 'BAD_REQUEST',
        statusCode: 400,
      );

  @override
  String toString() => message;
}

File: apps/api_server/src/errors/app_error.ts
export class AppError extends Error {
  constructor(
    public readonly statusCode: number,
    public readonly code: string,
    message: string,
  ) {
    super(message);
    this.name = 'AppError';
  }

  static notFound(resource: string): AppError {
    return new AppError(404, 'NOT_FOUND', `${resource} not found`);
  }

  static badRequest(message: string): AppError {
    return new AppError(400, 'BAD_REQUEST', message);
  }

  static unauthorized(message = 'Unauthorized'): AppError {
    return new AppError(401, 'UNAUTHORIZED', message);
  }

  static conflict(message: string): AppError {
    return new AppError(409, 'CONFLICT', message);
  }

  /**
   * A durably-recorded unresolved operation. Distinct from CONFLICT on purpose:
   * a conflict is retryable, this is not — the database, the target and the
   * projected profile disagree and a human has to look. Clients must be able to
   * tell them apart from a machine-readable code, not by matching prose.
   */
  static reconciliationRequired(message: string): AppError {
    return new AppError(409, 'RECONCILIATION_REQUIRED', message);
  }

  static forbidden(message = 'Forbidden'): AppError {
    return new AppError(403, 'FORBIDDEN', message);
  }

  static internal(message = 'Internal server error'): AppError {
    return new AppError(500, 'INTERNAL_ERROR', message);
  }
}
